include/xtl/xoptional.hpp

```diff
@@ -189,111 +189,111 @@ namespace xtl
 
         // Constructors
         inline xoptional()
-            : m_value(), m_flag(false)
+            : m_value(), m_flag(!EXISTS)
         {
         }
 
         template <class T,
           std::enable_if_t<
             conjunction<
-              negation<std::is_same<xoptional<CT, CB>, std::decay_t<T>>>,
+              negation<std::is_same<xoptional<CT, CB, EXISTS>, std::decay_t<T>>>,
               std::is_constructible<CT, T&&>,
               std::is_convertible<T&&, CT>
             >::value,
             bool
           > = true>
         inline constexpr xoptional(T&& rhs)
-            : m_value(std::forward<T>(rhs)), m_flag(true)
+            : m_value(std::forward<T>(rhs)), m_flag(EXISTS)
         {
         }
 
         template <class T,
           std::enable_if_t<
             conjunction<
-              negation<std::is_same<xoptional<CT, CB>, std::decay_t<T>>>,
+              negation<std::is_same<xoptional<CT, CB, EXISTS>, std::decay_t<T>>>,
               std::is_constructible<CT, T&&>,
               negation<std::is_convertible<T&&, CT>>
             >::value,
             bool
           > = false>
         inline explicit constexpr xoptional(T&& value)
-            : m_value(std::forward<T>(value)), m_flag(true)
+            : m_value(std::forward<T>(value)), m_flag(EXISTS)
         {
         }
 
-        template <class CTO, class CBO,
+        template <class CTO, class CBO, bool EXISTSO,
           std::enable_if_t<
             conjunction<
-              negation<std::is_same<xoptional<CT, CB>, xoptional<CTO, CBO>>>,
+              negation<std::is_same<xoptional<CT, CB, EXISTS>, xoptional<CTO, CBO, EXISTSO>>>,
               std::is_constructible<CT, std::add_lvalue_reference_t<std::add_const_t<CTO>>>,
               std::is_constructible<CB, std::add_lvalue_reference_t<std::add_const_t<CBO>>>,
               conjunction<
                 std::is_convertible<std::add_lvalue_reference_t<std::add_const_t<CTO>>, CT>,
                 std::is_convertible<std::add_lvalue_reference_t<std::add_const_t<CBO>>, CB>
               >,
-              negation<detail::converts_from_xoptional<CT, CTO, CBO>>
+              negation<detail::converts_from_xoptional<CT, CTO, CBO, EXISTSO>>
             >::value,
             bool
           > = true>
-        inline constexpr xoptional(const xoptional<CTO, CBO>& rhs)
-            : m_value(rhs.value()), m_flag(rhs.has_value())
+        inline constexpr xoptional(const xoptional<CTO, CBO, EXISTSO>& rhs)
+            : m_value(rhs.value()), m_flag(rhs.flag())
```

Comment:
`m_flag` probably cannot be initialized with `rhs.flag()` when `EXISTS` and `EXISTSO` are different.